bundles/umljava/tools.vitruv.applications.umljava.java2uml/src/tools/vitruv/applications/umljava/java2uml/JavaToUmlClassifier.reactions

```diff
@@ -375,15 +375,34 @@ reaction JavaPackageCreated {
     call {
     	detectOrCreateUmlModel(newValue)
 		registerPredefinedUmlPrimitiveTypes(correspondenceModel, newValue.eResource.resourceSet)
-	    createUmlPackage(newValue)
+	    createOrFindUmlPackage(newValue)
     }
 }
 
-routine createUmlPackage(java::Package jPackage) {
+routine createOrFindUmlPackage(java::Package jPackage) {
 	match {
         val uModel = retrieve uml::Model corresponding to UMLPackage.Literals.MODEL
 		require absence of uml::Package corresponding to jPackage
 	}
+	action {
+		call {
+			val uPackage = findUmlPackage(uModel, jPackage.name)
+			if(uPackage === null) {
+				createUmlPackage(jPackage, uModel)
+			} else {
+				addPackageCorrespondence(jPackage, uPackage)
+			}
+		}
+	}
+}
+
+routine addPackageCorrespondence(java::Package jPackage, uml::Package uPackage) {
+	action {
+		 add correspondence between uPackage and jPackage
+	}
+}
```

Comment:
This routine contains a pattern that might frequently appear during upcoming fixes:

```python
if model element absent:
    if can be found by name:
        add missing correspondence
    else:
        create missing model element
```
 
Every time the change propagation specification order is [PCM --> UML, PCM --> Java, ...] correlating model elements (UML & Java) are already created but are missing the correspondence. Right now this results in duplicated model elements, as it just creates a new model element. The pattern seen above tries to locate the already existing model element and adds the missing correspondence instead of creating a duplicate.

While this pull request fixes such an issue for packages, I already found another problem caused by a similar issue regarding interfaces. These issues will most likely arise for other model elements as well.

Reply:
This is indeed the pattern that may have to be applied at several places in the PCM-Java reactions implementations to deal with the fact that PCM-UML and UML-Java reactions already created the appropriate elements.